Kernel/FileSystem/VFSRootContext.cpp

```diff
@@ -116,32 +116,223 @@ ErrorOr<NonnullRefPtr<VFSRootContext>> VFSRootContext::create_with_empty_ramfs()
+{
+    m_details.with([](auto& details) {
+        details.attached_by_process.set();
+        details.attach_count++;
     });
 }
 
 bool VFSRootContext::mount_point_exists_at_custody(Custody& mount_point)
 {
-    return m_mounts.with([&](auto& mounts) -> bool {
-        return any_of(mounts, [&mount_point](auto const& existing_mount) {
+    return m_details.with([&](auto& details) -> bool {
+        return any_of(details.mounts, [&mount_point](auto const& existing_mount) {
             return existing_mount.host_custody() && VirtualFileSystem::check_matching_absolute_path_hierarchy(*existing_mount.host_custody(), mount_point);
         });
     });
 }
 
+ErrorOr<void> VFSRootContext::from_mount_for_host_custody(Custody const& current_custody, Function<void(Mount const&)> callback) const
+{
+    return m_details.with([&](auto const& details) -> ErrorOr<void> {
+        // NOTE: We either search for the root mount or for a mount that has a parent custody!
+        if (!current_custody.parent()) {
+            for (auto const& mount : details.mounts) {
+                if (!mount.host_custody()) {
+                    callback(mount);
+                    return {};
+                }
+            }
+            // NOTE: There must be a root mount entry, so fail if we don't find it.
+            VERIFY_NOT_REACHED();
+        } else {
+            for (auto const& mount : details.mounts) {
+                if (mount.host_custody() && VirtualFileSystem::check_matching_absolute_path_hierarchy(*mount.host_custody(), current_custody)) {
+                    callback(mount);
+                    return {};
+                }
+            }
+        }
+        return Error::from_errno(ENODEV);
+    });
+}
+
+ErrorOr<void> VFSRootContext::apply_to_mount_for_host_custody(Custody const& current_custody, Function<void(Mount&)> callback)
+{
+    return m_details.with([&](auto& details) -> ErrorOr<void> {
+        // NOTE: We either search for the root mount or for a mount that has a parent custody!
+        if (!current_custody.parent()) {
+            for (auto& mount : details.mounts) {
+                if (!mount.host_custody()) {
+                    TRY(VFSRootContext::validate_mount_not_immutable_while_being_used(details, mount));
+                    callback(mount);
+                    return {};
+                }
+            }
+            // NOTE: There must be a root mount entry, so fail if we don't find it.
+            VERIFY_NOT_REACHED();
+        } else {
+            for (auto& mount : details.mounts) {
+                if (mount.host_custody() && VirtualFileSystem::check_matching_absolute_path_hierarchy(*mount.host_custody(), current_custody)) {
+                    TRY(VFSRootContext::validate_mount_not_immutable_while_being_used(details, mount));
+                    callback(mount);
+                    return {};
+                }
+            }
+        }
+        return Error::from_errno(ENODEV);
+    });
+}
```

Comment:
These 2 functions are very similar, can we instead pass an argument whether or not it should verify that the mount is immutable and used by a process? Or maybe even do the check in the callback?

I also find these function names a bit confusing, but I guess the 2nd one was already called like that.

Reply:
You're probably right about merging the functions, but I don't have any better name (and as you said, it was like that before for the 2nd function anyway).
If you have suggestions then I think we should discuss them :)

Reply:
I merged the functions and made it hopefully more clear now :)